### PR TITLE
Thematic MNS

### PR DESCRIPTION
**Pull request by besserox**
We have this Thematic Module Naming Scheme that we are using on our clusters at the University of Luxembourg. It has been developed by @sylmarien. 

With this scheme, module are generated like *moduleclass/name/version*. For example:
```
biology/ABySS/1.3.4-goolf-1.4.10     compiler/GCC/4.9.2    mpi/MVAPICH2/2.0.1-GCC-4.8.4
```

Would you like to integrate this?
I'm not sure that this code is ready to be merged. Please let me know what needs to be fixed.

### CODE REVIEW

**Comment by boegel** [comment]
Jenkins: ok to test

File: easybuild/tools/module_naming_scheme/thematic_mns.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#        return os.path.join(ec['name'], det_full_ec_version(ec))
+
+    def is_short_modname_for(self, short_modname, name):
+        """
+        Determine whether the specified (short) module name is a module for software with the specified name.
+        Default implementation checks via a strict regex pattern, and assumes short module names are of the form:
+        <name>/<version>[-<toolchain>]
+        """
+        
+        modname_regex = re.compile('^\S+/%s/\S+$' % re.escape(name))
```
Comment (boegel) [inline easybuild/tools/module_naming_scheme/thematic_mns.py]: should be `^[^/]+/...` ?

File: easybuild/tools/module_naming_scheme/thematic_mns.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        res = bool(modname_regex.match(short_modname))
+
+        tup = (short_modname, name, modname_regex.pattern, res)
+        self.log.debug("Checking whether '%s' is a module name for software with name '%s' via regex %s: %s" % tup)
```
Comment (boegel) [inline easybuild/tools/module_naming_scheme/thematic_mns.py]: move this above the `match`?

File: easybuild/tools/module_naming_scheme/thematic_mns.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        
+        return os.path.join(ec['moduleclass'], ec['name'], det_full_ec_version(ec))
+
+# Maybe, can be usefull to still have the moduleclass when using module list (since it supports regex)
+#    def det_short_module_name(self, ec):
+#        """
+#        Determine short module name, i.e. the name under which modules will be exposed to users.
+#        Examples: GCC/4.8.3, OpenMPI/1.6.5, OpenBLAS/0.2.9, HPL/2.1, Python/2.7.5
+#        """
+#        return os.path.join(ec['name'], det_full_ec_version(ec))
```
Comment (boegel) [inline easybuild/tools/module_naming_scheme/thematic_mns.py]: this should be removed?

File: easybuild/tools/module_naming_scheme/thematic_mns.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import os
+import re
+
+from easybuild.tools.module_naming_scheme import ModuleNamingScheme
+from easybuild.tools.module_naming_scheme.utilities import det_full_ec_version
+
+class ThematicMNS(ModuleNamingScheme):
```
Comment (boegel) [inline easybuild/tools/module_naming_scheme/thematic_mns.py]: we already have `CategorizedHMNS` (HMNS for Hierarchical MNS), so `CategorizedMNS` would make more sense, but would be close enough to `CategorizedHMNS` to cause confusion...

it's unclear where the term "thematic" comes from though...

maybe `CategorizedModuleNamingScheme` ?

**Comment by boegel** [comment]
@besserox: having a test for this new mns would be nice... (cfr. `test_module_naming_scheme` in `test/framework/module_generator.py`

**Comment by besserox** [comment]
@boegel I fixed your comments. Can you look at it again?

**Comment by boegel** [comment]
Looks great, I see no reason no to include this, so going in.

Thanks @besserox!

**Pull request merged**